extensions/pooling/src/__tests__/acceptance/pooling.acceptance.ts

```diff
@@ -0,0 +1,370 @@
+    expect(poolService.pool.size).to.eql(0);
+    poolService.start();
+    const res = await poolService.acquire();
+    expect(res.status).to.eql('created');
+    expect(poolService.pool.available).to.eql(1);
+    expect(poolService.pool.borrowed).to.eql(1);
+    expect(poolService.pool.size).to.eql(2);
+    await poolService.release(res);
+    expect(poolService.pool.available).to.eql(2);
+    expect(poolService.pool.borrowed).to.eql(0);
+    expect(poolService.pool.size).to.eql(2);
+  });
+
+  it('honors poolOptions.max', async () => {
+    const poolService = await givenPoolService({
+      min: 1,
+      max: 2,
+      acquireTimeoutMillis: 100,
+    });
+    // 1st
+    const res1 = await poolService.acquire();
+    expect(poolService.pool.borrowed).to.eql(1);
+    // 2nd
+    await poolService.acquire();
+    expect(poolService.pool.available).to.eql(0);
+    expect(poolService.pool.borrowed).to.eql(2);
+    expect(poolService.pool.size).to.eql(2);
+    // 3rd has to wait
+    await expect(poolService.acquire()).to.be.rejectedWith(
+      /ResourceRequest timed out/,
+    );
+    await poolService.release(res1);
+    expect(poolService.pool.available).to.eql(1);
+    await poolService.acquire();
+    expect(poolService.pool.available).to.eql(0);
+    expect(poolService.pool.borrowed).to.eql(2);
+    expect(poolService.pool.size).to.eql(2);
+  });
+
+  it('destroys a resource from the pool', async () => {
+    const poolService = await givenPoolService();
+    const res = await poolService.acquire();
+    expect(res.status).to.eql('created');
+    await poolService.release(res);
+    expect(res.status).to.eql('created');
+    await poolService.stop();
+    expect(res.status).to.eql('destroyed');
+  });
+
+  it('validates a resource during acquire', async () => {
+    const poolService = await givenPoolService({
+      testOnBorrow: true,
+    });
+    const res = await poolService.acquire();
+    expect(res.status).to.eql('validated');
+  });
+
+  it('fails a resource during acquire', async () => {
+    const poolService = await givenPoolService(
+      {
+        testOnBorrow: true,
+      },
+      'invalid',
+    );
+    const res = await poolService.acquire();
+    expect(res.status).to.eql('validated');
+    // The first creation fails and generic-pool calls `create` again
+    expect(res.id).to.eql(2);
+  });
+
+  it.skip('validates a resource during release', async () => {
+    const poolService = await givenPoolService({
+      testOnReturn: true,
+    });
+    const res = await poolService.acquire();
+    expect(res.status).to.eql('created');
+    res.status = 'invalid';
+    await poolService.release(res);
+    expect(poolService.pool.isBorrowedResource(res)).to.be.false();
+  });
+
+  it('invokes resource-level acquire/release methods', async () => {
+    const poolService = await givenPoolService(
+      {max: 5},
+      undefined,
+      ExpensiveResourceWithHooks,
```

Comment:
Should the signature of `givenPoolService()` be changed to make this possible:

```
givenPoolService(
      {max: 5},
      ExpensiveResourceWithHooks
);
```

?